File: unirig-ui/backend/app/diagnostics.py
"""
Diagnostics and error reporting endpoint
"""
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from typing import List, Dict, Optional
from datetime import datetime
import os
import logging

logger = logging.getLogger(__name__)
router = APIRouter()


class ErrorDetails(BaseModel):
    code: str
    message: str
    stack: Optional[str] = None


class SystemInfo(BaseModel):
    platform: str
    language: str
    cookiesEnabled: bool
    onLine: bool


class Viewport(BaseModel):
    width: int
    height: int


class DiagnosticReport(BaseModel):
    timestamp: str
    userAgent: str
    viewport: Viewport
    sessionId: Optional[str]
    errorDetails: ErrorDetails
    recentActions: List[str]
    systemInfo: SystemInfo


@router.post('/diagnostics')
async def report_diagnostics(report: DiagnosticReport):
    """
    Receive and log diagnostic information from frontend
    
    Saves error reports to disk for debugging and analysis.
    """
    try:
        # Log to console
        logger.error('Frontend error report received:')
        logger.error(f"Session: {report.sessionId or 'N/A'}")
        logger.error(f"Error Code: {report.errorDetails.code}")
        logger.error(f"Error Message: {report.errorDetails.message}")
        logger.error(f"User Agent: {report.userAgent}")
        logger.error(f"Recent Actions: {report.recentActions}")
        
        # Save to diagnostics log file if configured
        log_dir = os.environ.get('DIAGNOSTICS_LOG_DIR', '/tmp/unirig-diagnostics')
        os.makedirs(log_dir, exist_ok=True)
        
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        session_id = report.sessionId or 'unknown'
        log_file = os.path.join(log_dir, f'error_{timestamp}_{session_id}.txt')
        
        with open(log_file, 'w') as f:
            f.write('UniRig Error Report\n')
            f.write('=' * 80 + '\n\n')
            f.write(f"Timestamp: {report.timestamp}\n")
            f.write(f"Session ID: {session_id}\n\n")
            
            f.write('Error Details:\n')
            f.write('-' * 80 + '\n')
            f.write(f"Code: {report.errorDetails.code}\n")
            f.write(f"Message: {report.errorDetails.message}\n")
            if report.errorDetails.stack:
                f.write(f"\nStack Trace:\n{report.errorDetails.stack}\n")
            
            f.write('\nSystem Information:\n')
            f.write('-' * 80 + '\n')
            f.write(f"User Agent: {report.userAgent}\n")
            f.write(f"Platform: {report.systemInfo.platform}\n")
            f.write(f"Language: {report.systemInfo.language}\n")
            f.write(f"Cookies Enabled: {report.systemInfo.cookiesEnabled}\n")
            f.write(f"Online: {report.systemInfo.onLine}\n")
            f.write(f"Viewport: {report.viewport.width}x{report.viewport.height}\n")
            
            f.write('\nRecent Actions:\n')
            f.write('-' * 80 + '\n')
            if report.recentActions:
                for i, action in enumerate(report.recentActions, 1):
                    f.write(f"{i}. {action}\n")
            else:
                f.write('None recorded\n')
        
        logger.info(f'Diagnostics saved to: {log_file}')
        
        return {
            'success': True,
            'message': 'Diagnostics received',
            'log_file': log_file
        }
        
    except Exception as e:
        logger.error(f'Failed to process diagnostics: {str(e)}')
        raise HTTPException(
            status_code=500,
            detail={
                'error': 'Failed to process diagnostics',
                'details': str(e)
            }
        )

File: unirig-ui/frontend/src/utils/diagnostics.ts
/**
 * Diagnostics and error reporting utilities
 */

export interface DiagnosticInfo {
  timestamp: string;
  userAgent: string;
  viewport: {
    width: number;
    height: number;
  };
  sessionId: string | null;
  errorDetails: {
    code: string;
    message: string;
    stack?: string;
  };
  recentActions: string[];
  systemInfo: {
    platform: string;
    language: string;
    cookiesEnabled: boolean;
    onLine: boolean;
  };
}

/**
 * Collect diagnostic information for error reporting
 */
export function collectDiagnostics(error: any, sessionId: string | null, recentActions: string[] = []): DiagnosticInfo {
  return {
    timestamp: new Date().toISOString(),
    userAgent: navigator.userAgent,
    viewport: {
      width: window.innerWidth,
      height: window.innerHeight,
    },
    sessionId,
    errorDetails: {
      code: error.code || error.name || 'UNKNOWN',
      message: error.message || String(error),
      stack: error.stack,
    },
    recentActions: recentActions.slice(-10), // Last 10 actions
    systemInfo: {
      platform: navigator.platform,
      language: navigator.language,
      cookiesEnabled: navigator.cookieEnabled,
      onLine: navigator.onLine,
    },
  };
}

/**
 * Format diagnostics for clipboard or download
 */
export function formatDiagnostics(diagnostics: DiagnosticInfo): string {
  return `UniRig Error Report
Generated: ${diagnostics.timestamp}

Session ID: ${diagnostics.sessionId || 'N/A'}

Error Details:
- Code: ${diagnostics.errorDetails.code}
- Message: ${diagnostics.errorDetails.message}

System Information:
- User Agent: ${diagnostics.userAgent}
- Platform: ${diagnostics.systemInfo.platform}
- Language: ${diagnostics.systemInfo.language}
- Viewport: ${diagnostics.viewport.width}x${diagnostics.viewport.height}
- Online: ${diagnostics.systemInfo.onLine ? 'Yes' : 'No'}

Recent Actions:
${diagnostics.recentActions.map((action, i) => `${i + 1}. ${action}`).join('\n') || 'None recorded'}

Stack Trace:
${diagnostics.errorDetails.stack || 'Not available'}
`;
}

/**
 * Copy diagnostics to clipboard
 */
export async function copyDiagnosticsToClipboard(diagnostics: DiagnosticInfo): Promise<boolean> {
  try {
    const formatted = formatDiagnostics(diagnostics);
    await navigator.clipboard.writeText(formatted);
    return true;
  } catch (error) {
    console.error('Failed to copy diagnostics:', error);
    return false;
  }
}

/**
 * Download diagnostics as a text file
 */
export function downloadDiagnostics(diagnostics: DiagnosticInfo): void {
  const formatted = formatDiagnostics(diagnostics);
  const blob = new Blob([formatted], { type: 'text/plain' });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = `unirig-error-${Date.now()}.txt`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}

/**
 * Send diagnostics to backend for logging (if endpoint exists)
 */
export async function reportDiagnostics(diagnostics: DiagnosticInfo): Promise<boolean> {
  try {
    const response = await fetch('/api/diagnostics', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(diagnostics),
    });
    return response.ok;
  } catch (error) {
    console.error('Failed to report diagnostics:', error);
    return false;
  }
}
